specs/core/1_shard-data-chains.md

```diff
@@ -75,207 +91,204 @@ The following types are defined, mapping into `DomainType` (little endian):
-        (SHARD_COUNT * TARGET_COMMITTEE_SIZE),
-        len(get_active_validator_indices(state, later_start_epoch)) //
-        (SHARD_COUNT * TARGET_COMMITTEE_SIZE),
-    ) + 1
-
-    index = slot % committee_count
-    earlier_committee = get_period_committee(state, earlier_start_epoch, shard, index, committee_count)
-    later_committee = get_period_committee(state, later_start_epoch, shard, index, committee_count)
+    earlier_committee = get_period_committee(state, get_period_start_epoch(epoch, lookback=2), shard)
+    later_committee = get_period_committee(state, get_period_start_epoch(epoch, lookback=1), shard)
 
     # Take not-yet-cycled-out validators from earlier committee and already-cycled-in validators from
     # later committee; return a sorted list of the union of the two, deduplicated
-    return sorted(list(set(
-        [i for i in earlier_committee if epoch % PERSISTENT_COMMITTEE_PERIOD < get_switchover_epoch(state, epoch, i)]
-        + [i for i in later_committee if epoch % PERSISTENT_COMMITTEE_PERIOD >= get_switchover_epoch(state, epoch, i)]
-    )))
+    return sorted(set(
+        [i for i in earlier_committee if epoch % EPOCHS_PER_SHARD_PERIOD < i % EPOCHS_PER_SHARD_PERIOD]
+        + [i for i in later_committee if epoch % EPOCHS_PER_SHARD_PERIOD >= i % EPOCHS_PER_SHARD_PERIOD]
+    ))
 ```
 
-### `get_shard_proposer_index`
+### `get_shard_block_proposer_index`
 
 ```python
-def get_shard_proposer_index(state: BeaconState,
-                             shard: Shard,
-                             slot: Slot) -> Optional[ValidatorIndex]:
+def get_shard_block_proposer_index(state: BeaconState,
+                                   shard: Shard,
+                                   slot: ShardSlot) -> Optional[ValidatorIndex]:
     # Randomly shift persistent committee
     persistent_committee = list(get_persistent_committee(state, shard, slot))
     seed = hash(state.current_shuffling_seed + int_to_bytes(shard, length=8) + int_to_bytes(slot, length=8))
-    random_index = bytes_to_int(seed[0:8]) % len(persistent_committee)
-    persistent_committee = persistent_committee[random_index:] + persistent_committee[:random_index]
-
-    # Search for an active proposer
-    for index in persistent_committee:
-        if is_active_validator(state.validators[index], get_current_epoch(state)):
-            return index
-
-    # No block can be proposed if no validator is active
-    return None
+    i = 0
+    while True:
+        candidate_index = persistent_committee[(slot + i) % len(persistent_committee)]
+        random_byte = hash(seed + int_to_bytes(i // 32, length=8))[i % 32]
+        effective_balance = state.validators[candidate_index].effective_balance
+        if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte:
+            if is_active_validator(state.validators[candidate_index], compute_epoch_of_shard_slot(slot)):
+                return ValidatorIndex(candidate_index)
+        i += 1
```

Comment:
Edge case: Is it possible that none of the candidates is active? (It was handled previously)

Reply:
So replace `while True:` with `for i in range(len(persistent_committee)):`, and then add a `return None` at the end?

Reply:
@vbuterin sounds reasonable. :)

Reply:
I'll handle this

Reply:
opted for an exit condition prior to the loop. We want the loop to have the quality that if balances are low it can take multiple passes through the committee to eventually probabilistically choose one.

Reply:
@djrtwo but ad30945b9b0cb63d00b0ea09fb0ccba9c6e35957 has an issue: it may return inactive validator. I think @vbuterin's suggestion is correct, and the client can *choose* to optimize it by checking if all the possible candidates are inactive before the while loop as you put.

Reply:
So we should exit if no active, and filter the validators for only active.
We want to be able to loop through the active validators more than once which @vbuterin's proposed solution doesn't give us

Reply:
I think this is what we want -- https://github.com/ethereum/eth2.0-specs/pull/1294/commits/21f4e0334b60e217c6f1ac7948265448ab8665c0

Reply:
Yep, looks good to me!